pkg/settings/setting.go

```diff
@@ -117,45 +117,106 @@ type NonMaskedSetting interface {
+	//     end-users) in CockroachCloud Serverless, AND a single value
+	//     must apply to all virtual clusters simultaneously;
+	//
+	//     (If this part of the condition does not hold, consider
+	//     ApplicationLevel instead.)
+	//
+	//   - AND, its value is never needed in the application layer (i.e.
+	//     it is not read from SQL code, HTTP handlers or other code
+	//     that would run in SQL pods in CC Serverless).
+	//
+	//     (If this part of the condition does not hold, consider
+	//     SystemVisible instead.)
 	SystemOnly Class = iota
 
-	// TenantReadOnly settings are visible to non-system tenants but cannot be
-	// modified by the tenant. Values for these settings are set from the system
-	// tenant and propagated from the host cluster.
-	TenantReadOnly
+	// SystemVisible settings are specific to the KV/storage layer and
+	// are also visible to virtual clusters.
+	//
+	// As a rule of thumb, use this class if:
+	//
+	//   - the setting may be accessed from the shared KV/storage layer;
+	//
+	//   - AND the setting should only be controllable by SREs (not
+	//     end-users) in CockroachCloud Serverless, AND a single value
+	//     must apply to all virtual clusters simultaneously;
+	//
+	//     (If this part of the condition does not hold, consider
+	//     ApplicationLevel instead.)
+	//
+	//   - AND, its value is sometimes needed in the application layer
+	//     (i.e. it may be read from SQL code, HTTP handlers or other
+	//     code that could run in SQL pods in CC Serverless).
+	//
+	//     (If this part of the condition does not hold, consider
+	//     SystemOnly instead.)
+	SystemVisible
 
-	// TenantWritable settings are visible to and can be modified by non-system
-	// tenants. The system can still override these settings; the overrides are
-	// propagated from the host cluster.
-	TenantWritable
+	// ApplicationLevel settings are readable and can optionally be
+	// modified by virtual clusters.
+	//
+	// As a rule of thumb, use this class if:
+	//
+	//   - the setting is never accessed by the shared KV/storage layer;
+	//
+	//   - AND, any of the following holds:
+	//
+	//     - end-users should legitimately be able to modify
+	//       the setting in CockroachCloud Serverless;
+	//
+	//       (If this part of the condition does not hold, but the other
+	//       part of the condition below does hold, consider still using
+	//       ApplicationLevel and force an override using ALTER VIRTUAL
+	//       CLUSTER SET CLUSTER SETTING. This makes the
+	//       ApplicationLevel setting effectively read-only from the
+	//       virtual cluster's perspective, because system overrides
+	//       cannot be modified by the virtual cluster.)
+	//
+	//     - OR, different virtual clusters should be able to
+	//       use different values for the setting.
+	//
+	//       (If this part of the condition does not hold, consider
+	//       SystemOnly or SystemVisible instead.)
```

Comment:
I wonder if we should provide some more guidance here. Namely, _most_ settings should be ApplicationLevel if they _aren't_ in the KV/storage layer.

Reply:
Done.